src/ScriptCs.Core/ScriptLibraryComposer.cs

```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Common.Logging;
+using ScriptCs.Contracts;
+using System.IO;
+
+namespace ScriptCs
+{
+    public class ScriptLibraryComposer : IScriptLibraryComposer
+    {
+        private readonly IFileSystem _fileSystem;
+        private readonly IFilePreProcessor _preProcessor;
+        private readonly IPackageContainer _packageContainer;
+        private readonly IPackageAssemblyResolver _packageAssemblyResolver;
+        private readonly ILog _logger;
+
+        public ScriptLibraryComposer(IFileSystem fileSystem, IFilePreProcessor preProcessor, IPackageContainer packageContainer, IPackageAssemblyResolver packageAssemblyResolver, ILog logger)
+        {
+            Guard.AgainstNullArgument("fileSystem", fileSystem);
+            Guard.AgainstNullArgument("preProcessor", preProcessor);
+            Guard.AgainstNullArgument("packageContainer", packageContainer);
+            Guard.AgainstNullArgument("packageAssemblyResolver", packageAssemblyResolver);
+            Guard.AgainstNullArgument("logger", logger);
+
+            _fileSystem = fileSystem;
+            _preProcessor = preProcessor;
+            _packageContainer = packageContainer;
+            _packageAssemblyResolver = packageAssemblyResolver;
+            _logger = logger;
+        }
+
+        internal string GetMainScript(IPackageObject package)
+        {
+            var script =
+                package.GetContentFiles()
+                    .SingleOrDefault(f => f.EndsWith("main.csx", StringComparison.InvariantCultureIgnoreCase));
+            
+            if (script != null)
+            {
+                _logger.DebugFormat("Found main script: {0}", script);    
+            }
+            
+            return script;
+        }
+
+        public virtual string ScriptLibrariesFile
+        {
+            get { return "ScriptLibraries.csx"; }
+        }
+
+        public void Compose(StringBuilder builder = null)
+        {
+            var namespaces = new List<string>();
+            var references = new List<string>();
+
+            var packagesPath = Path.Combine(_fileSystem.CurrentDirectory, _fileSystem.PackagesFolder);
+            var packageReferences = _packageAssemblyResolver.GetPackages(_fileSystem.CurrentDirectory);
+            var packageScriptsPath = Path.Combine(packagesPath, ScriptLibrariesFile);
+
+            if (!_fileSystem.DirectoryExists(packagesPath) || _fileSystem.FileExists(packageScriptsPath))
```

Comment:
Good catch on the delete. The refresh option should not be there at all. It
was supposed to be pulled in favor of -install having the logic.
On Fri, Feb 27, 2015 at 6:04 PM Paul Bouwer <notifications@github.com>
wrote:

> In src/ScriptCs.Core/ScriptLibraryComposer.cs
> <https://github.com/scriptcs/scriptcs/pull/928#discussion_r25552637>:
>
> > +
> > +        public virtual string ScriptLibrariesFile
> > +        {
> > +            get { return "ScriptLibraries.csx"; }
> > +        }
> > +
> > +        public void Compose(StringBuilder builder = null)
> > +        {
> > +            var namespaces = new List<string>();
> > +            var references = new List<string>();
> > +
> > +            var packagesPath = Path.Combine(_fileSystem.CurrentDirectory, _fileSystem.PackagesFolder);
> > +            var packageReferences = _packageAssemblyResolver.GetPackages(_fileSystem.CurrentDirectory);
> > +            var packageScriptsPath = Path.Combine(packagesPath, ScriptLibrariesFile);
> > +
> > +            if (!_fileSystem.DirectoryExists(packagesPath) || _fileSystem.FileExists(packageScriptsPath))
>
> I modified CalculatorMain.csx by adding .. to the logged message as
> follows:
>
> public double Add(double a, double b) {
>   logger.Info(String.Format("Adding .. {0} + {1}", a, b));  <==
>   return a+b;
> }
>
> Running it results in:
>
> c:\Dev > scriptcs CalcTest.csx -- -loglevel info
> [2015-02-28T11:39:35]  INFO  Adding .. 5 + 2
> 7
>
> I deleted the ScriptCs.Calculator.0.1.0 folder in the scriptcs_packages
> folder to force a re-install of the package.
>
> c:\Dev > scriptcs -install
> WARN: Cannot find: ScriptCs.Calculator 0.1.0.0
> Installing packages...
> Installed: ScriptCs.Calculator 0.1.0.0
> Package installation succeeded.
> Saving packages in scriptcs_packages.config...
> Updating scriptcs_packages.config...
> Skipped ScriptCs.Calculator because it already exists.
> Skipped ScriptCs.Contracts because it already exists.
> Skipped ScriptCs.Logger.ScriptPack because it already exists.
> Successfully updated scriptcs_packages.config.
>
> c:\Dev > scriptcs CalcTest.csx -- -loglevel info
> [2015-02-28T11:40:11]  INFO  Adding .. 5 + 2
> 7
>
> The function was not replaced with the original indicating that the
> ScriptLibraries.csx had not been deleted. @glennblock
> <https://github.com/glennblock> - If you check the line referenced in
> your note above you'll notice that the delete is being logged but not
> actioned.
>
> if (_fileSystem.DirectoryExists(packagesFolder))
> {
>     _logger.DebugFormat("Deleting: {0}", scriptLibrariesFile);
> }
>
> It also seems as though the RefreshScriptLibraries option is not wired
> up. I can't find any code that references RefreshScriptLibraries from
> src/ScriptCs/ScriptCsArgs.cs.
>
> Running the following does not refresh the cached Script Library.
>
> C:\Dev > scriptcs CalcTest.csx -RefreshScriptLibraries -- -loglevel info
> [2015-02-28T11:58:08]  INFO  Adding .. 5 + 2
> 7
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scriptcs/scriptcs/pull/928/files#r25552637>.
>